File: connected_graph.py
from collections import defaultdict

def is_connected( graph, start ):
    visited = set()
    def dfs( v ):
        visited.add( v )
        for neighbor in graph[v]:
            if neighbor not in visited:
                dfs( neighbor )
    dfs( start )
    return len( visited ) == len( graph )

graph = { 0: [1, 2], 1: [0, 2], 2: [0, 1] }
print( "Is graph connected?", is_connected( graph, 0 ) )

File: connected_graph.rb
class Graph
  def initialize
    @adj_list = Hash.new { |h, k| h[k] = [] }
  end

  def add_edge( u, v )
    @adj_list[u] << v
    @adj_list[v] << u
  end

  def connected?
    visited = {}
    dfs( 0, visited )
    visited.size == @adj_list.keys.size
  end
  
  private
  
  def dfs( node, visited )
    visited[node] = true
    @adj_list[node].each { |neighbor| dfs(neighbor, visited) unless visited[neighbor] }
  end
end

g = Graph.new
[ [0,1],[1,2],[2,3] ].each { |u,v| g.add_edge(u,v) }
puts "Is graph connected? #{g.connected?}"
